File: src/Parking/Payment.fs
module Payment

type CheckNumber = CheckNumber of int

type UpiId = UpiId of string

type CreditCardNumber = CreditCardNumber of int

type CardType = VISA | MASTER

type CreditCardInfo = {
    Number: CreditCardNumber
    Type: CardType
}

type PaymentMethod = 
    | Cash
    | Check of CheckNumber
    | UPI of UpiId
    | CreditCard of CreditCardInfo

type Currency = INR | USD

type PaymentAmount = PaymentAmount of decimal

type Payment = {
    Amount : PaymentAmount
    Currency : Currency
    Method : PaymentMethod
}

File: src/ParkingCSharp/Payment/Payment.cs
namespace ParkingCSharp.Payment
{
    public enum CardType { VISA, MASTER }

    public abstract class PaymentMethod { }

    public class CashPaymentMethod : PaymentMethod { }
    public class CheckPaymentMethod : PaymentMethod { public int CheckNumber { get; set; } }
    public class UPIPaymentMethod : PaymentMethod { public string Upi { get; set; } }
    public class CreditCardPaymentMethod : PaymentMethod
    {
        public int CreditCardNumber { get; set; }
        public CardType Type { get; set; }
    }

    public enum CurrencyType { INR, USD }


    public class Payment
    {
        public decimal Amount { get; set; }
        public CurrencyType Currency { get; set; }
        public PaymentMethod Method { get; set; }
    }
}
